File: src/es/um/josecefe/rueda/resolutor/ResolutorV8.java
/*
 * Copyright (C) 2016 José Ceferino Ortega
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package es.um.josecefe.rueda.resolutor;

import es.um.josecefe.rueda.modelo.AsignacionDia;
import es.um.josecefe.rueda.modelo.AsignacionDiaV5;
import es.um.josecefe.rueda.modelo.Dia;
import es.um.josecefe.rueda.modelo.Horario;

import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.toList;

/**
 * ResolutorV8
 * <p>
 * Versión paralela del resolutor -> No implementa LNV, sino que sigue una
 * estrategia no guiada, solo podada (la estimación es innecesaria)
 *
 * @author josecefe
 */
public class ResolutorV8 extends ResolutorAcotado {

    private final static boolean DEBUG = false;

    private final static boolean ESTADISTICAS = true;
    private final static int CADA_EXPANDIDOS_EST = 1000;
    private final static int CADA_MILIS_EST = 1000;

    private static final int PESO_COTA_INFERIOR_NUM_DEF = 1;
    private static final int PESO_COTA_INFERIOR_DEN_DEF = 2;
    private final ContextoResolucion contexto = new ContextoResolucion();
    private final EstadisticasV8 estGlobal = new EstadisticasV8();
    private Set<Horario> horarios;
    private Map<Dia, AsignacionDiaV5> solucionFinal;
    private int[] tamanosNivel;
    private double[] nPosiblesSoluciones;
    private double totalPosiblesSoluciones;
    private AtomicInteger cotaInferiorCorte;
    private Nodo RAIZ;
    private long ultMilisEst; // La ultima vez que se hizo estadística

    public ResolutorV8() {

    }

    private void inicializa() {
        continuar = true;

        contexto.inicializa(horarios);
        contexto.pesoCotaInferiorNum = PESO_COTA_INFERIOR_NUM_DEF;
        contexto.pesoCotaInferiorDen = PESO_COTA_INFERIOR_DEN_DEF;

        solucionFinal = null;

        if (ESTADISTICAS) {
            tamanosNivel = contexto.solucionesCandidatas.keySet().stream().mapToInt(k -> contexto.solucionesCandidatas.get(k).size()).toArray();
            totalPosiblesSoluciones = IntStream.of(tamanosNivel).mapToDouble(i -> (double) i).reduce(1.0, (a, b) -> a * b);
            if (DEBUG) {
                System.out.println("Nº de posibles soluciones: " + IntStream.of(tamanosNivel).mapToObj(Double::toString).collect(Collectors.joining(" * ")) + " = "
                        + totalPosiblesSoluciones);
            }
            double acum = 1.0;
            nPosiblesSoluciones = new double[tamanosNivel.length];

            for (int i = tamanosNivel.length - 1; i >= 0; i--) {
                nPosiblesSoluciones[i] = acum;
                acum = acum * tamanosNivel[i];
            }
        }
    }

    @Override
    public Map<Dia, ? extends AsignacionDia> resolver(Set<Horario> horarios) {
        return resolver(horarios, Integer.MAX_VALUE - 1);
    }

    @Override
    public Map<Dia, ? extends AsignacionDia> resolver(Set<Horario> horarios, int cotaInfCorteInicial) {
        this.horarios = horarios;
        if (horarios.isEmpty()) {
            return Collections.emptyMap();
        }

        if (ESTADISTICAS) {
            estGlobal.inicia();
        }

        inicializa();

        if (ESTADISTICAS) {
            estGlobal.setTotalPosiblesSoluciones(totalPosiblesSoluciones);
            estGlobal.actualizaProgreso();
            if (DEBUG) {
                System.out.format("Tiempo inicializar =%s\n", estGlobal.getTiempoString());
            }
        }

        // Preparamos el algoritmo
        RAIZ = new Nodo(contexto);
        Nodo actual = RAIZ;
        Nodo mejor = actual;
        cotaInferiorCorte = new AtomicInteger(cotaInfCorteInicial + 1); //Lo tomamos como cota superior

        mejor = branchAndBound(actual, mejor).orElse(mejor); //Cogemos el nuevo mejor

        //Estadisticas finales
        if (ESTADISTICAS) {
            estGlobal.setFitness(mejor.getCosteEstimado());
            estGlobal.actualizaProgreso();
            if (DEBUG) {
                System.out.println("====================");
                System.out.println("Estadísticas finales");
                System.out.println("====================");
                System.out.println(estGlobal);
                System.out.println("Solución final=" + mejor);
                System.out.println("-----------------------------------------------");
            }
        }
        // Construimos la solución final
        if (mejor.getCosteEstimado() < cotaInfCorteInicial) {
            solucionFinal = mejor.getSolucion();
        } else {
            solucionFinal = null;
        }

        return solucionFinal;
    }

    private Optional<Nodo> branchAndBound(Nodo actual, Nodo mejorPadre) {
        Nodo mejor = mejorPadre;
        if (ESTADISTICAS && estGlobal.incExpandidos() % CADA_EXPANDIDOS_EST == 0L && System.currentTimeMillis() - ultMilisEst > CADA_MILIS_EST) {
            ultMilisEst = System.currentTimeMillis();
            estGlobal.setFitness(cotaInferiorCorte.get());
            estGlobal.actualizaProgreso();
            if (DEBUG) {
                System.out.println(estGlobal);
                System.out.println("-- Trabajando con " + actual);
            }
        }
        if (actual.getCotaInferior() < cotaInferiorCorte.get() && continuar) { //Estrategia de poda: si la cotaInferior >= C no seguimos
            if (ESTADISTICAS) {
                estGlobal.addGenerados(tamanosNivel[actual.getIndiceDia() + 1]);
            }
            if (actual.getIndiceDia() + 2 == contexto.dias.length) { // Los hijos son terminales
                if (ESTADISTICAS) {
                    estGlobal.addTerminales(tamanosNivel[actual.getIndiceDia() + 1]);
                }
                Optional<Nodo> mejorHijo = actual.generaHijos(true).min(Nodo::compareTo); //true para paralelo

                if (mejorHijo.isPresent() && mejorHijo.get().compareTo(mejor) < 0) {
                    mejor = mejorHijo.get();
                    int cota;
                    do {
                        cota = cotaInferiorCorte.get();
                    }
                    while (mejor.getCosteEstimado() < cota && !cotaInferiorCorte.compareAndSet(cota, mejor.getCosteEstimado()));
                    if (mejor.getCosteEstimado() < cota) {
                        if (ESTADISTICAS) {
                            estGlobal.setFitness(mejor.getCosteEstimado());
                            estGlobal.actualizaProgreso();
                        }
                        if (DEBUG) {
                            System.out.format("$$$$ A partir del padre=%s\n    -> mejoramos con el hijo=%s\n", actual, mejor);
                            System.out.format("** Nuevo (nueva solución) C: Anterior=%,d, Nuevo=%,d\n", cota, mejor.getCosteEstimado());
                        }
                    }
                }
            } else { // Es un nodo intermedio
                List<Nodo> lNF = actual.generaHijos(true).filter(n -> n.getCotaInferior() < cotaInferiorCorte.get()).collect(toList()); //PARALELO poner true
                OptionalInt menorCotaSuperior = lNF.stream().mapToInt(Nodo::getCotaSuperior).min();
                if (menorCotaSuperior.isPresent()) {
                    int cota;
                    do {
                        cota = cotaInferiorCorte.get();
                    }
                    while (menorCotaSuperior.getAsInt() < cota && !cotaInferiorCorte.compareAndSet(cota, menorCotaSuperior.getAsInt()));
                    if (menorCotaSuperior.getAsInt() < cota) {
                        if (ESTADISTICAS) {
                            estGlobal.setFitness(menorCotaSuperior.getAsInt());
                            estGlobal.actualizaProgreso();
                        }
                        if (DEBUG) {
                            System.out.format("** Nuevo C: Anterior=%,d, Nuevo=%,d\n", cota, menorCotaSuperior.getAsInt());
                        }
                    }
                    lNF.removeIf(n -> n.getCotaInferior() >= cotaInferiorCorte.get()); //Recalculamos lNF
                    // Limpiamos la LNV
                }
                if (ESTADISTICAS) {
                    estGlobal.addDescartados((tamanosNivel[actual.getIndiceDia() + 1] - lNF.size()) * nPosiblesSoluciones[actual.getIndiceDia() + 1]);
                }
                final Nodo mejorAhora = mejor;
                Optional<Nodo> mejorHijo = lNF.parallelStream().sorted().map(n -> branchAndBound(n, mejorAhora)).filter(Optional<Nodo>::isPresent).map(Optional<Nodo>::get).min(Nodo::compareTo);
                if (mejorHijo.isPresent() && mejorHijo.get().compareTo(mejor) < 0) { // Tenemos un hijo que mejora
                    mejor = mejorHijo.get();
                }
            }
        } else if (ESTADISTICAS) {
            if (continuar) {
                estGlobal.addDescartados(nPosiblesSoluciones[actual.getIndiceDia()]);
            }
        }

        return Optional.ofNullable(mejor);
    }

    @Override
    public Map<Dia, ? extends AsignacionDia> getSolucionFinal() {
        return solucionFinal;
    }

    @Override
    public Estadisticas getEstadisticas() {
        return estGlobal;
    }

    @Override
    public void setEstrategia(Estrategia estrategia) {
        contexto.estrategia = estrategia;
    }
}

File: src/es/um/josecefe/rueda/resolutor/ResolutorV8.kt
/*
 * Copyright (c) 2016-2017. Jose Ceferino Ortega Carretero
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package es.um.josecefe.rueda.resolutor

import es.um.josecefe.rueda.modelo.AsignacionDia
import es.um.josecefe.rueda.modelo.Dia
import es.um.josecefe.rueda.modelo.Horario
import kotlinx.coroutines.*
import java.util.concurrent.atomic.AtomicInteger

/**
 * ResolutorV8
 *
 *
 * Versión paralela del resolutor -> No implementa LNV, sino que sigue una
 * estrategia no guiada, solo podada (la estimación es innecesaria)
 *
 * @author josecefe
 */
class ResolutorV8 : ResolutorAcotado() {
    private var contexto: ContextoResolucionHeuristico? = null
    private var tamanosNivel: IntArray? = null
    private var nPosiblesSoluciones: DoubleArray? = null
    private val estGlobal = EstadisticasV7()
    override var solucionFinal: Map<Dia, AsignacionDia> = emptyMap()
        private set
    private var cotaInferiorCorte: AtomicInteger? = null
    private var ultMilisEst: Long = 0 // La ultima vez que se hizo estadística

    override fun resolver(horarios: Set<Horario>): Map<Dia, AsignacionDia> {
        return resolver(horarios, Integer.MAX_VALUE - 1)
    }

    override fun resolver(horarios: Set<Horario>, cotaInfCorte: Int): Map<Dia, AsignacionDia> {
        if (horarios.isEmpty()) {
            return emptyMap()
        }

        if (ESTADISTICAS) {
            estGlobal.inicia()
        }

        continuar = true

        val contexto1 = ContextoResolucionHeuristico(horarios)
        contexto = contexto1
        contexto1.pesoCotaInferiorNum = PESO_COTA_INFERIOR_NUM_DEF
        contexto1.pesoCotaInferiorDen = PESO_COTA_INFERIOR_DEN_DEF

        solucionFinal = emptyMap()

        tamanosNivel = contexto1.ordenExploracionDias.map { i -> contexto1.solucionesCandidatas[contexto1.dias[i]]!!.size }.toIntArray()
        val totalPosiblesSoluciones = tamanosNivel!!.map { i -> i.toDouble() }.fold(1.0) { a, b -> a * b }
        nPosiblesSoluciones = DoubleArray(tamanosNivel!!.size)
        if (ESTADISTICAS) {
            if (DEBUG) {
                println("Nº de posibles soluciones: " + tamanosNivel!!.map { it.toDouble().toString() }.joinToString(" * ") + " = "
                        + totalPosiblesSoluciones)
            }
            var acum = 1.0

            for (i in tamanosNivel!!.indices.reversed()) {
                nPosiblesSoluciones!![i] = acum
                acum *= tamanosNivel!![i]
            }
        }


        if (ESTADISTICAS) {
            estGlobal.totalPosiblesSoluciones = totalPosiblesSoluciones
            estGlobal.actualizaProgreso()
            if (DEBUG) {
                System.out.format("Tiempo inicializar =%s\n", estGlobal.tiempoString)
            }
        }

        // Preparamos el algoritmo
        val actual = Nodo(contexto1)
        cotaInferiorCorte = AtomicInteger(cotaInfCorte + 1) //Lo tomamos como cota superior

        val mejor = runBlocking { branchAndBound(actual, actual) }

        //Estadisticas finales
        if (ESTADISTICAS) {
            estGlobal.fitness = mejor.costeEstimado
            estGlobal.actualizaProgreso()
            if (DEBUG) {
                println("====================")
                println("Estadísticas finales")
                println("====================")
                println(estGlobal)
                println("Solución final=" + mejor)
                println("-----------------------------------------------")
            }
        }
        // Construimos la solución final
        if (mejor.costeEstimado < cotaInfCorte) {
            solucionFinal = mejor.solucion
        } else {
            solucionFinal = emptyMap()
        }

        return solucionFinal
    }

    private suspend fun branchAndBound(actual: Nodo, mejorPadre: Nodo): Nodo {
        var mejor = mejorPadre
        if (ESTADISTICAS && estGlobal.incExpandidos() % CADA_EXPANDIDOS_EST == 0L && System.currentTimeMillis() - ultMilisEst > CADA_MILIS_EST) {
            ultMilisEst = System.currentTimeMillis()
            estGlobal.fitness = cotaInferiorCorte!!.get()
            estGlobal.actualizaProgreso()
            if (DEBUG) {
                println(estGlobal)
                println("-- Trabajando con " + actual)
            }
        }
        if (actual.cotaInferior < cotaInferiorCorte!!.get() && continuar) { //Estrategia de poda: si la cotaInferior >= C no seguimos
            if (ESTADISTICAS) {
                estGlobal.addGenerados(tamanosNivel!![actual.nivel + 1].toDouble())
            }
            if (actual.nivel + 2 == contexto!!.dias.size) { // Los hijos son terminales
                if (ESTADISTICAS) {
                    estGlobal.addTerminales(tamanosNivel!![actual.nivel + 1].toDouble())
                }
                val mejorHijo = actual.generaHijos().min()

                if (mejorHijo != null && mejorHijo < mejor) {
                    mejor = mejorHijo
                    var cota: Int
                    do {
                        cota = cotaInferiorCorte!!.get()
                    } while (mejor.costeEstimado < cota && !cotaInferiorCorte!!.compareAndSet(cota, mejor.costeEstimado))
                    if (mejor.costeEstimado < cota) {
                        if (ESTADISTICAS) {
                            estGlobal.fitness = mejor.costeEstimado
                            estGlobal.actualizaProgreso()
                        }
                        if (DEBUG) {
                            System.out.format("$$$$ A partir del padre=%s\n    -> mejoramos con el hijo=%s\n", actual, mejor)
                            System.out.format("** Nuevo (nueva solución) C: Anterior=%,d, Nuevo=%,d\n", cota, mejor.costeEstimado)
                        }
                    }
                }
            } else { // Es un nodo intermedio
                val lNF = actual.generaHijos().filter { n -> n.cotaInferior < cotaInferiorCorte!!.get() }.toMutableList()
                val menorCotaSuperior = lNF.map { it.cotaSuperior }.min()
                if (menorCotaSuperior != null) {
                    var cota: Int
                    do {
                        cota = cotaInferiorCorte!!.get()
                    } while (menorCotaSuperior < cota && !cotaInferiorCorte!!.compareAndSet(cota, menorCotaSuperior))
                    if (menorCotaSuperior < cota) {
                        if (ESTADISTICAS) {
                            estGlobal.fitness = menorCotaSuperior
                            estGlobal.actualizaProgreso()
                        }
                        if (DEBUG) {
                            System.out.format("** Nuevo C: Anterior=%,d, Nuevo=%,d\n", cota, menorCotaSuperior)
                        }
                    }
                    lNF.removeAll { it.cotaInferior >= cotaInferiorCorte!!.get() } //Recalculamos lNF
                    // Limpiamos la LNV
                }
                if (ESTADISTICAS) {
                    estGlobal.addDescartados((tamanosNivel!![actual.nivel + 1] - lNF.size) * nPosiblesSoluciones!![actual.nivel + 1])
                }
                lNF.sort()
                val mejorHijo = if (lNF.size > 1 && actual.nivel + 3 < contexto!!.dias.size) { // Los hijos estan lejos de ser terminales)
                    val mejorHijoDef: MutableList<Deferred<Nodo>> = ArrayList<Deferred<Nodo>>(lNF.size)
                    //println("Lanzando una corutina en nivel ${actual.nivel} con ${lNF.size} hilos")
                    for (n in lNF) {
                        mejorHijoDef.add(GlobalScope.async{ branchAndBound(n, mejor) })
                    }

                    mejorHijoDef.map { it.await() }.min()
                } else {
                    lNF.map { branchAndBound(it, mejor) }.min()
                }
                if (mejorHijo != null && mejorHijo < mejor) { // Tenemos un hijo que mejora
                    mejor = mejorHijo
                }
            }
        } else if (ESTADISTICAS) {
            if (continuar) {
                estGlobal.addDescartados(nPosiblesSoluciones!![actual.nivel])
            }
        }

        return mejor //Devolvemos el mejor hijo
    }

    override val estadisticas: Estadisticas
        get() = estGlobal

    override var estrategia = Estrategia.EQUILIBRADO

    companion object {

        private const val DEBUG = false

        //private const val PARALELO = false

        private const val ESTADISTICAS = true
        private const val CADA_EXPANDIDOS_EST = 1000
        private const val CADA_MILIS_EST = 1000

        private const val PESO_COTA_INFERIOR_NUM_DEF = 1
        private const val PESO_COTA_INFERIOR_DEN_DEF = 2
    }
}
